sdk/src/Core/Amazon.Runtime/Credentials/Internal/_bcl45+netstandard/SsoTokenCache.cs

```diff
@@ -0,0 +1,164 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#if !NETSTANDARD13
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+using Amazon.Runtime.Internal.Util;
+using Amazon.Util;
+
+namespace Amazon.Runtime.Credentials.Internal
+{
+    /// <summary>
+    /// Cached Storage IO for SSO Tokens (<see cref="SsoToken"/>)
+    /// </summary>
+    public class SsoTokenCache
+    {
+        private readonly Logger _logger = Logger.GetLogger(typeof(SsoTokenCache));
+
+        public string StartUrl { get; }
+
+        private readonly string _cachePath;
+
+        /// <summary>
+        /// Class constructor
+        /// Caches SSO Tokens to disk under (user profile)/.aws/sso/cache/
+        /// </summary>
+        /// <param name="startUrl">The configured sso_start_url for the profile that credentials are being resolved for.</param>
+        public SsoTokenCache(string startUrl)
+            : this(
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".aws", "sso", "cache"),
+                startUrl)
+        {
+        }
+
+        /// <summary>
+        /// Overload used for unit testing (allows for custom storage location)
+        /// </summary>
+        /// <param name="cacheFolder">The folder containing Cached SSO Tokens</param>
+        /// <param name="startUrl">The configured sso_start_url for the profile that credentials are being resolved for.</param>
+        public SsoTokenCache(string cacheFolder, string startUrl)
+        {
+            StartUrl = startUrl;
+            
+            var filename = GetCacheFilename(StartUrl);
+            _cachePath = Path.Combine(cacheFolder, filename);
+        }
+
+        /// <summary>
+        /// Retrieves the SSO Token's AccessToken.
+        /// </summary>
+        /// <returns>
+        /// SSO Token AccessToken.
+        /// Returns null if the token has expired, or if the cache is invalid or does not exist.
+        /// </returns>
+        public string GetAccessToken()
+        {
+            var cache = Load();
+
+            if (cache.IsExpired())
+            {
+                return null;
+            }
+
+            return cache.AccessToken;
+        }
+
+        /// <summary>
+        /// Caches an SSO Token to disk.
+        /// </summary>
+        /// <param name="token">SSO Token to cache</param>
+        public bool TrySave(SsoToken token)
+        {
+            try
+            {
+                if (token.StartUrl != StartUrl)
```

Comment:
Since the spec requires the file to be loaded when the access token is needed, multiple credentials objects making reference to the same starturl would access the same file. When the cache reports a non-expired token, things would proceed smoothly for multiple credentials objects. You could get a scenario where multiple credentials objects are each told the token is expired, and each would attempt to perform the SSO flow, then write back a token. The only concern here would be if those writes occurred at the same time. Thoughts?